### ISSUE TITLE
stopping a jail with zfs-mounts takes long time

### ISSUE THREAD

**Issue by uropro** [cbsd/cbsd#439]
/usr/jails/jails-fstab/fstab.myjail.local
```
/usr/jails/jails-data/myjail-data/mnt /mnt nullfs rw 0 0
zssd/samba/Projects /mnt/Projects zfs rw 0 0
```
The first line creates the read-writeable mounting point.
`cbsd jstop myjail` produces the following output:
```
root@myserver:~ # cbsd jstop myjail
Skip for interface=auto: NIC specified as part of ip_addr string
persist mode:
Stoping jail: myjail, parallel timeout=0
.
Terminated
myjail: removed
invalid jail id or name
usage:
        unjail <jailid|jailname> <filesystem>

For the property list, run: zfs set|get

For the delegated permission list, run: zfs allow|unallow
Updating: pkg info
jstop done in 3 minutes and 3 seconds
```
After 3 minutes, the zfs dataset is unmounted , "mountpoint" is none and "jailed" is off, that is correct. But why the error above and why does it take 3 minutes?

FreeBSD version 12.0-RELEASE-p10
CBSD version ( cbsd version ): 12.1.1_1

**Comment by uropro** [comment]
The error occurs, because the command `zfs unjail myjail zssd/samba/Projects` is executed after terminating the jail... Shouldn't it be executed before terminating the jail?

**Comment by olevole** [comment]
Yes, that seems to be the reason. I believe that if unmounting is done before the jail termination, we need tools to verify the processes that may hold the file system (lsof, fstat ..) or use 'timeout' for 'zfs unjail' command. I have no way to debug this right now ;-(

**Comment by mekanix** [comment]
Could we make unmount in 2 phases? First phase for ZFS mounts after all processes die, second when jail is destroyed. We already have second phase, so the question is only where to put ZFS unmount. But yeah, to be clear, is it even possible with our architecture?

**Comment by olevole** [comment]
I cannot reproduce this behavior. Can someone write step-by-step instructions to reproduce ?

**Comment by olevole** [comment]
close, please reopen if you can reproduce this ( with step-by-step howto from clean CBSD installation ) with CBSD 12.1.3 (latest release at 2020-01-04). 

I tried with CBSD 12.1.3:

1) create new jail: `myjail`
2) zfs create zroot/ROOT/samba
3) 
```
cat > /usr/jails/jails-fstab/fstab.myjail.local <<EOF
/usr/jails/jails-data/myjail-data/mnt /mnt nullfs rw 0 0
zroot/ROOT/samba /mnt/Projects zfs rw 0 0
EOF
```

4) cbsd jset allow_zfs=1 jname=myjail
5) cbsd jstart myjail
6) cbsd jexec myjail

simulate any activity on the **/mnt/Projects/**
```
cd /mnt/Projects
nohup dd if=/dev/zero of=test.dat
```

7) In parallel/another session:
```
cbsd jstop myjail
```
in my case the container stops without any issues in 5 seconds

**Comment by uropro** [comment]
Hi Oleg! Sorry for the laaaaaate response, I have been totally occupied by other stuff. And also missing a test server. But that has changed now, at least the latter! :-)

Regarding this issue: If I follow your steps, I get the same behavior, i.e. quick stop.
I observe two things:
1. on starting (cbsd jstart myjail), it looks like that the jail is started twice:
```
root@tm2:~ # cbsd jstart test
Default NIC automatically selected: em0
set resource limit: [ ]
Starting jail: test, parallel timeout=5
test: created
ELF ldconfig path: /lib /usr/lib /usr/lib/compat /usr/local/lib
32-bit compatibility ldconfig path: /usr/lib32
Setting hostname: test.hq.triamec.com.
Creating and/or trimming log files.
Starting syslogd.
Clearing /tmp.
Updating motd:.
Starting cron.

Sat Feb 22 15:46:24 CET 2020
attachzfs: attached: zroot/ROOT/test
ELF ldconfig path: /lib /usr/lib /usr/lib/compat /usr/local/lib
32-bit compatibility ldconfig path: /usr/lib32
Setting hostname: test.hq.triamec.com.
Creating and/or trimming log files.
Starting syslogd.
Clearing /tmp.
Updating motd:.
Starting cron.

Sat Feb 22 15:46:25 CET 2020
jstart done in 3 seconds
```
Why this? Or is it just a repetition of the log? "normal",i.e. non-zfs,  jails do not show this behavior.

2. Having a second zfs mount below the first (zroot/ROOT/samba/home for example) will change things a bit: The starting log to the console has an additional line, at the very end:
```
cannot unmount '/mnt/samba': Device busy
```
However, this has no other impact. `cbsd jstop myjail` still works quickly.
Who is trying to do an unmount in this phase?
Again, no harm is done, another cosmetic thing.

3. The error message changed (now cbsd 12.1.5, FreeBSD 12.1) as compared to my original post (on my samba server): 
```
Terminated
myjail: removed
attachzfs: jail is not running: myjail
jail: myjail: /bin/sh -c /usr/local/bin/cbsd detachzfs fstab=/usr/jails/jails-fstab/fstab.myjail.local jname=myjail; : failed
Updating: pkg info
jstop done in 3 minutes and 2 seconds
```
The long pause takes place after "Terminated". Why does cbsd try to do attachzfs after destroying the jail?

4. Now I can reproduce the long (>3min) shutdown. Install samba in your myjail.
During the shutdown phase (cbsd jstop myjail) I can login into myjail. Inside myjail I have seen that there were still some processes running (samba in my case). Killing them let to immediate shutdown of the jail and the rest of the log appeared as listed above, except duration was then shorter than 3 minutes.
It looks like the shutdown of samba does not kill all of its processes. When samba is auto-started at start-up of the jail (rc.conf: samba_server_enable="YES"), even manually stopping the services (service samba_server stop) does not kill all samba processes. Manually killing them and restarting the samba server and then stopping again will stop all its processes... It seems that something is different right at startup... what could that be?

5. Enough for today! These issues are all unimportant, just a little annoying. Otherwise cbsd does a great job on our servers and I like to thank you for the great work!   -Urs

**Comment by uropro** [comment]
Point 1 above is causing real troubles: I see crontab, syslog and other enabled services started twice inside the jail. This is not desirable. Shall I file a separate issue for this?

**Comment by olevole** [comment]
reopen, investigation...

**Comment by Stefar77** [comment]
> reopen, investigation...

https://github.com/Stefar77/cbsd/commit/85bc9463ade34d1820a0028c405ee5472e68712f

**Comment by olevole** [comment]
>  even manually stopping the services (service samba_server stop) does not kill all samba processes. 


I think this is the key to the problem. You will get the same behavior if you do:

`nohup service samba_server start'

inside jail ( or any other service, with nohup ).

Unfortunately, in this case we can do little from CBSD side:

CBSD waits for `stop_timeout` seconds (by the way, you can change value for samba jail) before hard jail kill. This timeout is often necessary to give time to heavy processes to stop correctly (e.g. DB, noSQL ).

I think a good idea is to contact the samba maintainer on FreeBSD and describe the problem of not being able to correctly complete the samba process.

**Comment by olevole** [comment]
Close because the problem is caused by incorrect behavior of third-party software inside the jail.

**Comment by uropro** [comment]
I do not agree with your conclusion: Having a zfs attached inside a jail causes the jail being started twice, as you can see from the output under point 1. in my post end of February. The original issue  "stopping takes long time" is caused by the circumstance that certain services (such as samba) or all get started twice. Then, using `service samba stop` will only stop those started at the first or second starting, I don't know which one, probably the later. The other processes will not be stopped and that is the reason cbsd is waiting for the timeout.

Without zfs attach inside a jail everything works fine...

Having two zfs attaches causes the note under point 2.

**Comment by olevole** [comment]
I apologize - the double start of services was corrected in this commit: https://github.com/cbsd/cbsd/commit/f8a142b3d979800d0cbd234863575a93e34e9bc0, I mentioned this task in commit. Such confusion arises when many problems are reported in a single task. It’s better to create one task per problem. I think in a week I will make a new release

**Comment by uropro** [comment]
Great, thank you very much!!

**Comment by olevole** [comment]
@uropro: commited to ports tree https://svnweb.freebsd.org/ports?view=revision&revision=r529081
